commit c7647686da5fe6bc5369a983d845f82beca9c276
Author: Gordon Sim <gsim@redhat.com>
Date:   Tue May 15 19:59:58 2018 +0100

    PROTON-1845: treat attach after detach as starting new link

diff --git a/proton-c/src/core/transport.c b/proton-c/src/core/transport.c
index 96b54f2..094a017 100644
--- a/proton-c/src/core/transport.c
+++ b/proton-c/src/core/transport.c
@@ -1263,9 +1263,9 @@ pn_link_t *pn_find_link(pn_session_t *ssn, pn_bytes_t name, bool is_sender)
     if (link->endpoint.type == type &&
         // This function is used to locate the link object for an
         // incoming attach. If a link object of the same name is found
-        // which is closed both locally and remotely, assume that is
-        // no longer in use.
-        !((link->endpoint.state & PN_LOCAL_CLOSED) && (link->endpoint.state & PN_REMOTE_CLOSED)) &&
+        // which is remotely closed or detached, assume that is
+        // no longer in use and a new link is intended.
+        (!(link->endpoint.state & PN_REMOTE_CLOSED) && ((int32_t) link->state.remote_handle != -2)) &&
         pn_bytes_equal(name, pn_string_bytes(link->name)))
     {
       return link;
